### ISSUE TITLE
[BUG] [Wireserver Exception] 'latin-1' codec can't encode character '\u2018' in position 545

### ISSUE THREAD

**Issue by clund-nlg** [Azure/WALinuxAgent#2132]
Errors are continuously logged by WALinuxAgent like this:

2021-01-12T17:59:18.614253Z ERROR SendTelemetryHandler ExtHandler Event: name=WALinuxAgent, op=ReportEventErrors, message=DroppedEventsCount: 1
Reasons (first 5 errors): [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2018' in position 545: Body ('\u2018') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.: Traceback (most recent call last):
  File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 554, in call_wireserver
    resp = http_req(*args, **kwargs)
  File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 507, in http_post
    retry_delay=retry_delay)
  File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 420, in http_request
    redact_data=redact_data)
  File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 332, in _http_request
    conn.request(method=method, url=url, body=data, headers=headers)
  File "/usr/lib64/python3.6/http/client.py", line 1254, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib64/python3.6/http/client.py", line 1299, in _send_request
    body = _encode(body, 'body')
  File "/usr/lib64/python3.6/http/client.py", line 171, in _encode
    (name.title(), data[err.start:err.end], name)) from None
UnicodeEncodeError: 'latin-1' codec can't encode character '\u2018' in position 545: Body ('\u2018') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1087, in _send_event
    self.send_event(provider_id, buf[provider_id])
  File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1071, in send_event
    resp = self.call_wireserver(restutil.http_post, uri, data, header)
  File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 568, in call_wireserver
    ustr(e)))
azurelinuxagent.common.exception.ProtocolError: [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2018' in position 545: Body ('\u2018') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
, duration=0

This is new RHEL8 VM:
Operating system Linux (redhat 8.3)
Publisher RedHat
Offer RHEL
Plan 8-LVM

# waagent --version
WALinuxAgent-2.2.46 running on redhat 8.3
Python: 3.6.8
Goal state agent: 2.2.53

**Comment by mrc0mmand** [comment]
Can reproduce this as well in GitHub Actions simply by running a job:

```
Jan 13 14:46:09 fv-az193-685 provisioner[690]: [14:46:09] info: Provisioner.Framework.JobRunner[0]
Jan 13 14:46:09 fv-az193-685 provisioner[690]:       => [RequestId: 104948865][OrchestrationId: 39338b55-6393-4e20-8184-07a3ec799151.build.gcc_asan_ubsan]
Jan 13 14:46:09 fv-az193-685 provisioner[690]:       Job Process Monitor is scheduled to run in 300 seconds
Jan 13 14:46:11 fv-az193-685 python3[951]: 2021-01-13T14:46:11.432085Z ERROR SendTelemetryHandler ExtHandler Event: name=WALinuxAgent, op=ReportEventErrors, message=DroppedEventsCount: 1
Jan 13 14:46:11 fv-az193-685 python3[951]: Reasons (first 5 errors): [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2026' in position 546: Body ('\u2026') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.: Traceback (most recent call last):
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 554, in call_wireserver
Jan 13 14:46:11 fv-az193-685 python3[951]:     resp = http_req(*args, **kwargs)
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 502, in http_post
Jan 13 14:46:11 fv-az193-685 python3[951]:     return http_request("POST",
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 411, in http_request
Jan 13 14:46:11 fv-az193-685 python3[951]:     resp = _http_request(method,
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 332, in _http_request
Jan 13 14:46:11 fv-az193-685 python3[951]:     conn.request(method=method, url=url, body=data, headers=headers)
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "/usr/lib/python3.8/http/client.py", line 1255, in request
Jan 13 14:46:11 fv-az193-685 python3[951]:     self._send_request(method, url, body, headers, encode_chunked)
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "/usr/lib/python3.8/http/client.py", line 1300, in _send_request
Jan 13 14:46:11 fv-az193-685 python3[951]:     body = _encode(body, 'body')
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "/usr/lib/python3.8/http/client.py", line 164, in _encode
Jan 13 14:46:11 fv-az193-685 python3[951]:     raise UnicodeEncodeError(
Jan 13 14:46:11 fv-az193-685 python3[951]: UnicodeEncodeError: 'latin-1' codec can't encode character '\u2026' in position 546: Body ('\u2026') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
Jan 13 14:46:11 fv-az193-685 python3[951]: During handling of the above exception, another exception occurred:
Jan 13 14:46:11 fv-az193-685 python3[951]: Traceback (most recent call last):
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1087, in _send_event
Jan 13 14:46:11 fv-az193-685 python3[951]:     self.send_event(provider_id, buf[provider_id])
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1071, in send_event
Jan 13 14:46:11 fv-az193-685 python3[951]:     resp = self.call_wireserver(restutil.http_post, uri, data, header)
Jan 13 14:46:11 fv-az193-685 python3[951]:   File "bin/WALinuxAgent-2.2.53-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 567, in call_wireserver
Jan 13 14:46:11 fv-az193-685 python3[951]:     raise ProtocolError("[Wireserver Exception] {0}".format(
Jan 13 14:46:11 fv-az193-685 python3[951]: azurelinuxagent.common.exception.ProtocolError: [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2026' in position 546: Body ('\u2026') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
```

OS Info:
```
Current runner version: '2.275.1'
Operating System
  Ubuntu
  20.04.1
  LTS
Virtual Environment
  Environment: ubuntu-20.04
  Version: 20201210.0
  Included Software: https://github.com/actions/virtual-environments/blob/ubuntu20/20201210.0/images/linux/Ubuntu2004-README.md
```

**Comment by larohra** [comment]
Thanks for opening this issue @clund-nlg. 

This is a known issue which was actually already fixed in #1937. Though due to some compatibility issues with some other components on the VM, we had to revert it out #2067.

That being said, we are working on bringing that PR back soon. I will keep this issue open until then and resolve it whenever we re-merge the PR.

Thanks!

**Comment by robfordww** [comment]
Still no fix on this log-filler bug?  I have the same issue with Ubuntu-20.04

**Comment by larohra** [comment]
@robfordww this error shouldn't really fill your logs unless the buggy event is being generated periodically by some extension. What extensions do you have on the box? If possible try removing them as a quick mitigation maybe?

**Comment by robfordww** [comment]
@larohra Seem that I have VMAccessForLinux, OmsAgentForLinux, and LinuxAgent.AzureSecurityCenter; though latter says "Provisioning failed"

```
python3[1022]: 2021-04-09T13:13:21.471552Z ERROR SendTelemetryHandler ExtHandler Event: name=WALinuxAgent, op=ReportEventErrors, message=DroppedEventsCount: 1
 python3[1022]: Reasons (first 5 errors): [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2192' in position 547: Body ('\u2192') is >...... etc
```

approximately every one minute.

**Comment by larohra** [comment]
@robfordww from the snippet it does look like one of the extensions is producing some message periodically that needs UTF-8 encoding. As mentioned before, this is a known issue and we have a fix ready for it, we're just waiting for confirmation from the other components to ensure this change wont break them.

As a mitigation you can try removing the extensions if not needed. If they are needed, maybe try uninstalling and then installing it back, hopefully that can get rid of that recurring event. 

Sorry couldn't be of more help!

**Comment by ThomasPiskol** [comment]
Uninstalling all these extensions does not help. Even with no extension installed the error will show up continuously every minute:

```syslog
May 20 08:52:04 python3[1104]: 2021-05-20T08:52:04.754981Z ERROR SendTelemetryHandler ExtHandler Event: name=WALinuxAgent, op=ReportEventErrors, message=DroppedEventsCount: 1
May 20 08:52:04 python3[1104]: Reasons (first 5 errors): [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2192' in position 547: Body ('\u2192') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.: Traceback (most recent call last):
May 20 08:52:04 python3[1104]:   File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 571, in call_wireserver
May 20 08:52:04 python3[1104]:     resp = http_req(*args, **kwargs)
May 20 08:52:04 python3[1104]:   File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 517, in http_post
May 20 08:52:04 python3[1104]:     retry_delay=retry_delay)
May 20 08:52:04 python3[1104]:   File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 424, in http_request
May 20 08:52:04 python3[1104]:     redact_data=redact_data)
May 20 08:52:04 python3[1104]:   File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 334, in _http_request
May 20 08:52:04 python3[1104]:     conn.request(method=method, url=url, body=data, headers=headers)
May 20 08:52:04 python3[1104]:   File "/usr/lib/python3.6/http/client.py", line 1281, in request
May 20 08:52:04 python3[1104]:     self._send_request(method, url, body, headers, encode_chunked)
May 20 08:52:04 python3[1104]:   File "/usr/lib/python3.6/http/client.py", line 1326, in _send_request
May 20 08:52:04 python3[1104]:     body = _encode(body, 'body')
May 20 08:52:04 python3[1104]:   File "/usr/lib/python3.6/http/client.py", line 175, in _encode
May 20 08:52:04 python3[1104]:     (name.title(), data[err.start:err.end], name)) from None
May 20 08:52:04 python3[1104]: UnicodeEncodeError: 'latin-1' codec can't encode character '\u2192' in position 547: Body ('\u2192') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
May 20 08:52:04 python3[1104]: During handling of the above exception, another exception occurred:
May 20 08:52:04 python3[1104]: Traceback (most recent call last):
May 20 08:52:04 python3[1104]:   File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1129, in _send_event
May 20 08:52:04 python3[1104]:     self.send_event(provider_id, buf[provider_id])
May 20 08:52:04 python3[1104]:   File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1113, in send_event
May 20 08:52:04 python3[1104]:     resp = self.call_wireserver(restutil.http_post, uri, data, header)
May 20 08:52:04 python3[1104]:   File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 585, in call_wireserver
May 20 08:52:04 python3[1104]:     ustr(e)))
May 20 08:52:04 python3[1104]: azurelinuxagent.common.exception.ProtocolError: [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2192' in position 547: Body ('\u2192') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
May 20 08:52:04 python3[1104]: , duration=0
```

**Comment by larohra** [comment]
@ThomasPiskol can you turn on verbose logging and restart the agent and send me the `/var/log/waagent.log` file after the restart if you see that error again? It will help me narrow down where the bad event is coming from. 
Thanks!

**Comment by indgarg** [comment]
Is the issue solved? WALinuxAgent seems to be gone crazy. I am getting same error with below configuration 

WALinuxAgent-2.2.45 running on ubuntu 18.04
Python: 3.6.9
Goal state agent: 2.2.54.2

**The error from waagent.log**
2021-05-23T06:25:26.472357Z ERROR SendTelemetryHandler ExtHandler Event: name=WALinuxAgent, op=ReportEventErrors, message=DroppedEventsCount: 1
Reasons (first 5 errors): [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2192' in position 13292: Body ('\u2192') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.: Traceback (most recent call last):
  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 571, in call_wireserver
    resp = http_req(*args, **kwargs)
  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 517, in http_post
    retry_delay=retry_delay)
  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 424, in http_request
    redact_data=redact_data)
  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 334, in _http_request
    conn.request(method=method, url=url, body=data, headers=headers)
  File "/usr/lib/python3.6/http/client.py", line 1281, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1326, in _send_request
    body = _encode(body, 'body')
  File "/usr/lib/python3.6/http/client.py", line 175, in _encode
    (name.title(), data[err.start:err.end], name)) from None

**Comment by johanburati** [comment]
> This is a known issue which was actually already fixed in #1937. Though due to some compatibility issues with some other components on the VM, we had to revert it out #2067.
> 
> That being said, we are working on bringing that PR back soon. I will keep this issue open until then and resolve it whenever we re-merge the PR.
> 

It's been 5 months since the above comment, any idea when the PR will be re-merged ?

**Comment by larohra** [comment]
> It's been 5 months since the above comment, any idea when the PR will be re-merged ?

@johanburati sorry for the delay, we have a release going on right now with some big changes so we didn't want to lump this in with that. We're planning on adding this back in the next release. Will keep this thread posted.

**Comment by larohra** [comment]
>But I am seeing below lines I my syslog

>2021-05-30T16:45:11.258478Z INFO MonitorHandler ExtHandler Stopped tracking cgroup: Qualys.LinuxAgent.AzureSecurityCenter-1.0.0.15, path: /sys/fs/cgroup/cpu,cpuacct/system.slice/Qualys.LinuxAgent.AzureSecurityCenter_1.0.0.15_ad7c59f8-7b42-4c51-a34a-225d042ff6ee.scope
2021-05-30T16:45:11.258716Z INFO MonitorHandler ExtHandler Stopped tracking cgroup: Qualys.LinuxAgent.AzureSecurityCenter-1.0.0.15, path: /sys/fs/cgroup/memory/system.slice/Qualys.LinuxAgent.AzureSecurityCenter_1.0.0.15_ad7c59f8-7b42-4c51-a34a-225d042ff6ee.scope

@indgarg these messages are just benign logs which we log when we stock tracking an extension's cgroup. That being said, these logs should not have made it to the syslog at all, they should've been restricted to the waagent.log only. Could you please share the waagent.log and the syslog and I can take a look as how that happened. Thanks!

**Comment by g-spencer** [comment]
Azure VM with RedHat 8.4-LVM /var/log/messages still being spammed every minute with these messages:-

"Jun  7 14:58:53 ... python3.6[1864]: 2021-06-07T13:58:53.252516Z ERROR SendTelemetryHandler ExtHandler Event: name=WALinuxAgent, op=ReportEventErrors, message=DroppedEventsCount: 1
Jun  7 14:58:53 ... python3.6[1864]: Reasons (first 5 errors): [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2192' in position 547: Body ('\u2192') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.: Traceback (most recent call last):
Jun  7 14:58:53 ... python3.6[1864]:  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 571, in call_wireserver
Jun  7 14:58:53 ... python3.6[1864]:    resp = http_req(*args, **kwargs)
Jun  7 14:58:53 ... python3.6[1864]:  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 517, in http_post
Jun  7 14:58:53 ... python3.6[1864]:    retry_delay=retry_delay)
Jun  7 14:58:53 ... python3.6[1864]:  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 424, in http_request
Jun  7 14:58:53 ... python3.6[1864]:    redact_data=redact_data)
Jun  7 14:58:53 ... python3.6[1864]:  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/utils/restutil.py", line 334, in _http_request
Jun  7 14:58:53 ... python3.6[1864]:    conn.request(method=method, url=url, body=data, headers=headers)
Jun  7 14:58:53 ... python3.6[1864]:  File "/usr/lib64/python3.6/http/client.py", line 1269, in request
Jun  7 14:58:53 ... python3.6[1864]:    self._send_request(method, url, body, headers, encode_chunked)
Jun  7 14:58:53 ... python3.6[1864]:  File "/usr/lib64/python3.6/http/client.py", line 1314, in _send_request
Jun  7 14:58:53 ... python3.6[1864]:    body = _encode(body, 'body')
Jun  7 14:58:53 ... python3.6[1864]:  File "/usr/lib64/python3.6/http/client.py", line 175, in _encode
Jun  7 14:58:53 ... python3.6[1864]:    (name.title(), data[err.start:err.end], name)) from None
Jun  7 14:58:53 ... python3.6[1864]: UnicodeEncodeError: 'latin-1' codec can't encode character '\u2192' in position 547: Body ('\u2192') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
Jun  7 14:58:53 ... python3.6[1864]: During handling of the above exception, another exception occurred:
Jun  7 14:58:53 ... python3.6[1864]: Traceback (most recent call last):
Jun  7 14:58:53 ... python3.6[1864]:  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1129, in _send_event
Jun  7 14:58:53 ... python3.6[1864]:    self.send_event(provider_id, buf[provider_id])
Jun  7 14:58:53 ... python3.6[1864]:  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 1113, in send_event
Jun  7 14:58:53 ... python3.6[1864]:    resp = self.call_wireserver(restutil.http_post, uri, data, header)
Jun  7 14:58:53 ... python3.6[1864]:  File "bin/WALinuxAgent-2.2.54.2-py2.7.egg/azurelinuxagent/common/protocol/wire.py", line 585, in call_wireserver
Jun  7 14:58:53 ... python3.6[1864]:    ustr(e)))
Jun  7 14:58:53 ... python3.6[1864]: azurelinuxagent.common.exception.ProtocolError: [ProtocolError] [Wireserver Exception] 'latin-1' codec can't encode character '\u2192' in position 547: Body ('\u2192') is not valid Latin-1. Use body.encode('utf-8') if you want to send it encoded in UTF-8.
Jun  7 14:58:53 ... python3.6[1864]: , duration=0"

corresponding alerts are also logged in /var/log/waagent.log also every minute.
both files much larger than we'd like them to be, too much noise makes spotting genuine issues more difficult to spot.


installed versions:-

"waagent --version
WALinuxAgent-2.2.49.2 running on redhat 8.4
Python: 3.6.8
Goal state agent: 2.2.54.2"


Azure VM extensions installed and provisioning succeeded for:
AzurePolicyforLinux   -   Microsoft.GuestConfiguration.ConfigurationforLinux
LinuxAgent.AzureSecurityCenter   -   Qualys.LinuxAgent.AzureSecurityCenter
OMSExtension    -   Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux

**Comment by dzsibi** [comment]
Okay, so I went looking for a root cause. I looked at the actual message that is being encoded that contains the Unicode character that causes this exception, and found the following error message:
```
[ExtensionOperationError] Non-zero exit code: 127, /var/lib/waagent/Microsoft.Azure.Diagnostics.LinuxDiagnostic-3.0.131/diagnostic.py -install
[stdout]


[stderr]
Running scope as unit: Microsoft.Azure.Diagnostics.LinuxDiagnostic_3.0.131_46667e6e-43b2-408b-8f52-d7202dbbb92d.scope
/usr/bin/env: ‘python’: No such file or directory
```
The latin-1 encoding chokes on the `‘` character `env` returns. The actual issue is that there is no Python 2 on the VM, so no `python` executable is present. So I installed the `python2-minimal` package and added a symlink to make sure that `python` points to `python2` (this is for Ubuntu 20.04, but the solution should be similar on other systems):
```
sudo apt install python2-minimal
cd /usr/bin
sudo ln -s python2 python
```
After restarting walinuxagent, the installer could run and perform the update, and the error messages are now gone.

TL;DR **Make sure that `python` is in `PATH` and points to a valid installation of Python 2.**

**Comment by larohra** [comment]
Thanks for the deep dive @dzsibi, definitely helpful. For this though I just wanted to add some context, its the LinuxDiagnostic extension that has a dependency on `python2`, the agent itself can run on both python2 and python3. Not sure if the other users in this thread all have the same dependency on the LinuxDiagnostic extension.

**Comment by btorsaldivar** [comment]
Following the steps from @dzsibi to get a minimal python2 interpreter solved the issue for me. In my case, the error specifically mentioned the missing Python as the problem, this is for the `Microsoft.OSTCExtensions.DSCForLinux` extension

```
[ExtensionOperationError] Non-zero exit code: 127, /var/lib/waagent/Microsoft.OSTCExtensions.DSCForLinux-2.71.1.0/dsc.py -install [stdout] [stderr] Running scope as unit: Microsoft.OSTCExtensions.DSCForLinux_2.71.1.0_0afc384c-5b9a-4220-9480-30aab73ebb4e.scope /usr/bin/env: ‘python’: No such file or directory
```

**Comment by larohra** [comment]
Fixed in #2278

**Comment by g-spencer** [comment]
that appears more a workaround than a fix.
there is no 'python2-minimal' (nor similar) for RedHat anyway, so that has only 'fixed' it for Ubuntu

**Comment by dzsibi** [comment]
I think that the two problems are distinct:
- Installing Python 2 (which is no longer maintained and really shouldn't be needed) is a solution to the problem that generates these error messages - Microsoft.OSTCExtensions.DSCForLinux would have to be ported to Python 3 for this to be resolved, since most modern distros don't really want to provide Python 2 anymore. (You can try `sudo yum install python2` on RedHat, as they [promise](https://access.redhat.com/solutions/4455511) to keep it maintained until 2024.).
- The PR that resolved this issue fixes the encoding related problem, switching form latin-1 to utf-8, which makes the exception thrown by http/client.py disappear (in this case and in any of the other cases where there are non-latin characters in the error messages).

**Comment by larohra** [comment]
@dzsibi as you mentioned there are 2 different problems. The PR fixes the problem in the Guest Agent, where we were masking the real problem with some latin-1 encoding error message. 

For the other problem with py3 compatibility, that is something outside the scope of the Guest Agent. I would suggest opening up an issue on the DSCForLinux Extension's repo and have them triage it further.

Thanks!
